processing/src/test/java/org/apache/druid/java/util/metrics/StubServiceEmitter.java

```diff
@@ -68,15 +71,15 @@ public void emit(Event event)
    */
   public List<Event> getEvents()
   {
-    return events;
+    return new ArrayList<>(events);
   }
 
   /**
    * Gets all the metric events emitted since the previous {@link #flush()}.
    *
    * @return Map from metric name to list of events emitted for that metric.
    */
-  public Map<String, List<ServiceMetricEventSnapshot>> getMetricEvents()
+  public Map<String, Queue<ServiceMetricEventSnapshot>> getMetricEvents()
   {
     return metricEvents;
```

Comment:
Nit: I guess we can later update this method to also return a map containing read-only lists rather than the `Queue` or maybe just update the method to accept a metric name.

Reply:
yes; some enhancements could be made there - actually in the latchingemitter even after this patch an assimptotically `n^2` computation remained (over time)

I already had to undo a jfr fix in this patch as it was causing some other issues - and I'm afraid this won't fully solve the problem ; sometimes I could still reproduce the issue...with a lower probability....but its still there.

seems like recently it happens on apache/master more frequently - I have to collect more details

Reply:
Must be a bug then, because every `WaitCondition` tracks a `processedUntil` index.
And only processes new items after that.

Reply:
I guess now it might become `n^2` though since we are making a copy of `events`.

Reply:
If that is the case, we should expose another method in `StubServiceEmitter` to access the original events queue so that a copy is not needed. Alternatively, we can just make `events` field protected so that `LatchableEmitter` can access it.

Reply:
Yeah, I would have liked it to be something like that. But it would make the embedded tests (where the `LatchableEmitter` is actually used) much less readable.